core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala

```diff
@@ -654,6 +654,64 @@ class PlaintextConsumerTest extends BaseConsumerTest {
     assertEquals(7, consumer.committed(Set(tp2).asJava).get(tp2).offset)
   }
 
+  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumAndGroupProtocolNames)
+  @MethodSource(Array("getTestQuorumAndGroupProtocolParametersAll"))
+  def testCommitAsyncCompletedConsumerCloses(quorum: String, groupProtocol: String): Unit = {
+    // This is testing the contract that asynchronous offset commit are completed before the consumer
+    // is closed.
```

Comment:
I would extend with something like : ...before the consumer is closed, **even when no commit sync is performed as part of the close (due to auto-commit disabled, or simply because there no consumed offsets).**

That's the key as I see it, fixed in this PR, and being tested here. If the call to consumer.close performs a commit sync (needs auto-commit enabled and non-empty consumed offsets), then the async callbacks were always called I expect. The contract was not being respected in case the commit sync did not happen for some of the reasons mentioned above.

Reply:
Agreed!